### DISCUSSION TITLE
Forward declaration for enum

### DISCUSSION THREAD

**Discussion by AF9036** [nlohmann/json#4069] [Ideas]
When working with json_fwd.h I don't see easy way to forward declare num types serialization.

I had to create 2 macros:

//this one used in header file
```
#define FORWARD_DECLARATION_NLOHMANN_ENUM(ENUM_TYPE)         \
  template <typename BasicJsonType>                          \
  void to_json(BasicJsonType& j, const ENUM_TYPE& e);        \
  template <typename BasicJsonType>                          \
  void from_json(const BasicJsonType& j, ENUM_TYPE& e);      \
```

//this one used in cpp file together with NLOHMANN_JSON_SERIALIZE_ENUM
```
#define INSTANTIATE_NLOHMANN_ENUM(ENUM_TYPE)                 \
template void to_json<nlohmann::json>(nlohmann::json& j, const ENUM_TYPE& e);   \
template void from_json<nlohmann::json>(const nlohmann::json& j, ENUM_TYPE& e); \
```

Would it be possible to add FORWARD_DECLARATION_NLOHMANN_ENUM into json_fwd.hpp ? Or is there any other simpler way to forward declare those num serialization functions ?

**Comment by nlohmann** [comment]
This gap is real: json_fwd.hpp only forward-declares `basic_json`/`adl_serializer`, it does not provide a helper for forward-declaring an enum's `to_json`/`from_json` template overloads the way NLOHMANN_JSON_SERIALIZE_ENUM defines them. The two hand-rolled macros in this proposal are a reasonable, self-contained pattern; adding an official `NLOHMANN_JSON_SERIALIZE_ENUM_DECLARE`/`_DEFINE` pair to reduce header/source coupling for enum serialization would be a natural complement to the existing macro, though it's a fairly niche need most users work around easily on their own.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
